AFKR-11: Configure a patient list to show patients Waiting Room

mupagasi opened the pull request:
Added Global property called Waiting Room with an SQL query to filter patient with/without vital but no other obs registered on him

mupagasi committed:
[AFKR-11] added global property for waiting room patient list query

mupagasi committed:
[AFKR-11] fix global property for waiting room patient list query

rbuisson commented on configuration/globalproperties/additional_clinical_gp.xml:
```diff
@@ -26,5 +26,46 @@
+        where v.date_stopped is null AND v.voided = 0
+
+        AND v.patient_id IN (
+        select o.person_id from obs o where o.obs_id in (select obs_group_id from obs, concept_set cs where obs.concept_id = cs.concept_id AND cs.concept_set=6240) and o.concept_id=6240
```
Using the concept ID won't be a reliable way getting the vitals because this may change from one DB to another. We must use the UUID. I know it makes it more complicated :/
This applies for all concept IDs in the query.

mupagasi replied:
It is reasonable to use UUID let me fix these two issues

rbuisson reviewed (changes requested):
Thanks @mupagasi .
Few changes required.
- We can't use concept IDs in the query, as they are database dependent. Only the UUID is a reliable source. For instance, when I spin up my dev server, the Vitals concept ID is 6238. So let's use UUIDs: `773d36e9-119e-452c-91ee-4ac90245e504`
- The query does not fully work. It correctly displays patients with an active visit and no obs, or vitals/or not (once I change the concept ID as explained above). But when I close the visit and reopen a new one (so no obs on this new visit), the patient does not show in the tab

Step to reproduce:
- Create a patient and record one non-vitals obs: let's say ANC Referal: Yes
- Confirm the patient does not show in the Waiting Room, which is good.
- Close the visit.
- Open a new visit and do not record any obs.
- See that the patient does not show in the Waiting Room list, which is not good.

Possible reason:
- do you always ensure that you are filtering on obs on the **active** visit and not something else, such as today's observations.

mupagasi committed:
[FIX] AFKR-11: Configure a patient list to show patients Waiting Room

mupagasi commented:
- Removed the UUID usage in the query
- Fixed the issue that get raised after close and reopen a visit
- Improved the query to check all concept under vitals form (We have some limitation of the version of mysql, from version under 8 , we can't recursively traverse the concept_set table looking for concept_id under vitals form concept), here i used repetitive JOIN on concept_set table itself.

rbuisson commented on configuration/globalproperties/additional_clinical_gp.xml:
```diff
@@ -26,5 +26,72 @@
+				LEFT JOIN concept_set c4 ON c4.concept_id=c3.concept_set
+				WHERE (SELECT concept_id FROM openmrs.concept WHERE uuid='773d36e9-119e-452c-91ee-4ac90245e504') in (c2.concept_id,c3.concept_id,c4.concept_id)
+			    OR (SELECT concept_id FROM openmrs.concept WHERE uuid='80ef57a5-0d0c-48de-acdd-19e026c890b5') in (c1.concept_set) 
+			    OR (SELECT concept_id FROM openmrs.concept WHERE uuid='2c5bbbbf-f34d-4590-b624-9f17b75b7f00') in (c1.concept_set)
```
Which concept does this UUID refer to?

mupagasi replied:
ID,         NAME,               UUID
'3148',   'BMI STATUS',   '80ef57a5-0d0c-48de-acdd-19e026c890b5'
'3149',   'BMI',                  '2c5bbbbf-f34d-4590-b624-9f17b75b7f00'
'6240',   'Vital Signs',        '773d36e9-119e-452c-91ee-4ac90245e504'

rbuisson replied:
Thanks

rbuisson commented on configuration/globalproperties/additional_clinical_gp.xml:
```diff
@@ -26,5 +26,72 @@
+				LEFT JOIN concept_set c3 ON c3.concept_id=c2.concept_set
+				LEFT JOIN concept_set c4 ON c4.concept_id=c3.concept_set
+				WHERE (SELECT concept_id FROM openmrs.concept WHERE uuid='773d36e9-119e-452c-91ee-4ac90245e504') in (c2.concept_id,c3.concept_id,c4.concept_id)
+			    OR (SELECT concept_id FROM openmrs.concept WHERE uuid='80ef57a5-0d0c-48de-acdd-19e026c890b5') in (c1.concept_set) 
```
Which concept does this UUID refer to?

mupagasi replied:
ID,         NAME,               UUID
'3148',   'BMI STATUS',   '80ef57a5-0d0c-48de-acdd-19e026c890b5'
'6240',   'Vital Signs',        '773d36e9-119e-452c-91ee-4ac90245e504'

mupagasi replied:
I have used these 2 other concept BMI STATUS AND BMI as they are created automatically in obs table when the height and weight are recorded in the Vital Signs obs form and they don't belong to child concepts of vitals signs form concept in concept_set table

rbuisson replied:
Thanks.

rbuisson reviewed (commented):
Works good @mupagasi 👍 
I have left 2 comments on the UUID you have used, wondering what concept they refer to.
